src/convert/impls.rs

```diff
@@ -294,12 +294,20 @@ impl<T> OptionIntoWasmAbi for NonNull<T> {
     }
 }
 
-impl<T> FromWasmAbi for Option<NonNull<T>> {
+impl<T> FromWasmAbi for NonNull<T> {
     type Abi = u32;
 
     #[inline]
     unsafe fn from_abi(js: Self::Abi) -> Self {
-        NonNull::new(js as *mut T)
+        // SAFETY: Checked in bindings.
```

Comment:
I couldn't actually find any user-facing documentation, do you know of any? Otherwise I'm going to add it in this PR as well.

I also wouldn't know any other unsoundness in this regard besides `char` and `NonNull`.

Reply:
Oh wait, I don't think we shouldn't be following what `char` does, it's considered a bug: #2269.

So then, I'm not really sure what the rules are around what happens when JS passes in an invalid value when `--debug` is off. If it's completely the wrong type it seems to either coerce it to some garbage-but-sound value, or crash. So then I guess the issue with this and `char` is that they can cause issues without actually crashing? I don't really know.

Anyway, in light of that I think either that check should be included even without `--debug` or a plain `NonNull` shouldn't be allowed in the first place.

If we were to go the check route, the actual type checking should probably be kept to `--debug`. When `--debug` is off, it should just be a check that, after coercion, it isn't null (possibly within Rust), since as I mentioned `wasm-bindgen`'s current approach to incorrect types without `--debug` is to be garbage-but-sound.

What was the reasoning for adding `NonNull` support in the first place? This is starting to feel like an overcomplicated way of automatically inserting a `NonNull::new(ptr).unwrap()`.

Reply:
> Anyway, in light of that I think either that check should be included even without `--debug` or a plain `NonNull` shouldn't be allowed in the first place.
>
> If we were to go the check route, the actual type checking should probably be kept to `--debug`. When `--debug` is off, it should just be a check that, after coercion, it isn't null (possibly within Rust), since as I mentioned `wasm-bindgen`'s current approach to incorrect types without `--debug` is to be garbage-but-sound.

I'm more leaning towards crashing.

In Rust we have a similar scenario with integer overflow, integer overflow checks are disabled without debug mode for performance reasons, but dividing by zero panics. My interpretation here is that if we are doing the check anyway, we opt for correctness while we are at it as we don't really gain speed by coercing.

Though I'm not sure how much sense the performance argument makes considering we are doing through the FFI boundary. But I still prefer correctness as I don't really see the upside in coercing at this point.

> What was the reasoning for adding `NonNull` support in the first place? This is starting to feel like an overcomplicated way of automatically inserting a `NonNull::new(ptr).unwrap()`.

Personally I don't have one, but after #3852 I realized that my original reason for not implementing it didn't make much sense anymore in light that so many other types are just converted to garbage or in the case of `char` are unsound.

Again, apologies for the lack of context in OP!

---

Let me know what you think.

Reply:
> My interpretation here is that if we are doing the check anyway, we opt for correctness while we are at it as we don't really gain speed by coercing.

My thinking was that I wanted it to be consistent with how invalid types get handled during other conversions without `--debug`, but seeing as the behaviour's undefined anyway (it breaks if you enable `--debug`!) it doesn't particularly matter. I agree there isn't a performance concern.

Reply:
Done!
Will do a follow-up PR for `char`.